ELM cleanup unused data type OpenAcc prep

peterdschwartz opened the pull request:
These changes are made to prepare for openACC high-resolution runs and fix issues that 
are particularly difficult for GPUs to handle.  All the unused data types have been removed
from ELM subroutines (e.g., carbonstate/flux_vars).

A separate module timeinfoMod  was created that has the relevant time variables (time step, etc..)
 as global variables.  

dynConsBiogeochemMod::dyn_cnbal_patch and dynSubgridAdjustmentsMod were significantly
refactored.  The dynamic memory allocation is detrimental for GPU performance. 

[BFB]

rgknox commented:
I'm still looking through the changes, but my first impression is that there is a ton of great code cleaning in here.  Thanks @peterdschwartz

thorntonpe commented on components/elm/src/cpl/lnd_import_export.F90:
```diff
@@ -1145,7 +1145,7 @@ subroutine lnd_import( bounds, x2l, atm2lnd_vars, glc2lnd_vars)
          co2_ppmv_val = co2_ppmv
        else if (co2_type_idx == 1) then               ! CO2 prognostic, value from coupler field
-         co2_ppmv_val = x2l(index_x2l_Sa_co2prog,i)
+         co2_ppmv_val = 0d0 !x2l(index_x2l_Sa_co2prog,i)
```
Why is the coupler field for prognostic CO2 replaced by 0?

peterdschwartz replied:
I have changed this back.

thorntonpe reviewed (commented):
I've looked through all the changes. Most of this I've had a chance to see in previous code reviews, and I have just a few minor comments /questions to address, posted previously.

rgknox commented on components/elm/src/biogeochem/AllocationMod.F90:
```diff
@@ -62,55 +55,74 @@ module AllocationMod
   !
   ! !PRIVATE DATA MEMBERS:
-  real(r8)              :: dt                   !decomp timestep (seconds)
+  !!real(r8)              :: dt                   !decomp timestep (seconds)
```
remove?

peterdschwartz replied:
removed

rgknox commented on components/elm/src/biogeochem/AllocationMod.F90:
```diff
@@ -283,36 +305,33 @@ subroutine AllocationInit ( bounds, elm_fates)
-    call cnallocate_carbonphosphorus_only_set(carbonphosphorus_only)
+     !$acc update device(carbon_only, carbonnitrogen_only,&
+     !$acc carbonphosphorus_only)
+    !!call cnallocate_carbon_only_set(carbon_only)
```
remove?

peterdschwartz replied:
removed.  And also went ahead and removed the use statements for the "cnallocate_*" functions

rgknox commented on components/elm/src/biogeochem/AllocationMod.F90:
```diff
@@ -370,326 +356,135 @@ subroutine Allocation1_PlantNPDemand (bounds, num_soilc, filter_soilc, num_soilp
-      if (spinup_state == 1 .and. yr .gt. nyears_ad_carbon_only) then 
-        call cnallocate_carbon_only_set(.false.)
+      if (spinup_state == 1 .and. yr .gt. nyears_ad_carbon_only) then
+         carbon_only = .false.
```
if this is being set to false, do we also need to set CN or CNP to true? Or check them?

peterdschwartz replied:
I believe carbon_only is the only one adjusted mid-run after initialization based on the namelist.  Having that if statement being checked every time step is inelegant but I don't know what would have to change to get rid of that.

rgknox commented on components/elm/src/biogeochem/AllocationMod.F90:
```diff
@@ -1070,24 +865,16 @@ end subroutine Allocation1_PlantNPDemand
-                            nitrogenstate_vars, nitrogenflux_vars           , &
-                            phosphorusstate_vars,phosphorusflux_vars        , &
-                            soilstate_vars,waterstate_vars, &
-                            elm_fates)
```
@peterdschwartz I added this argument (elm_fates) in PR #3990.  We don't use this argument now, but in my next set of changes I will need it.  Although I suppose I can just add it back in again, I will probably work on top of this PR or its merged result depending on timing of things.

peterdschwartz replied:
The way I have been handling elm_fates arguments is to place a 'use elm_instMod , only : alm_fates'   at the top of those modules.   This makes it easy to wrap any fates subroutine calls in  #ifndef _OPENACC flags.   

Do you think this is an acceptable solution or does it lose needed functionality for fates?  if so, then I can revert those changes back to passing alm_fates as an argument.

rgknox replied:
ok, this should work. We will just need to pass in the thread index then.

rgknox commented on components/elm/src/biogeochem/AllocationMod.F90:
```diff
@@ -1368,7 +1101,7 @@ subroutine Allocation2_ResolveNPLimit (bounds, num_soilc, filter_soilc  , &
                      actual_immob_vr(c,j) = potential_immob_vr(c,j)
                      sminn_to_plant_vr(c,j) = col_plant_ndemand(c) * nuptake_prof(c,j)
-                  else if ( cnallocate_carbon_only() .or. cnallocate_carbonphosphorus_only() ) then !.or. &
+                  else if ( carbon_only .or. carbonphosphorus_only ) then !.or. &
```
I like setting these to logicals rather than functions, more efficient

peterdschwartz commented on components/elm/src/biogeophys/PhotosynthesisMod.F90:
```diff
@@ -3746,8 +3710,8 @@ subroutine spacF(p,c,x,f,qflx_sun,qflx_sha, &
-            tsai(p)*params_inst%kmax(veg_pp%itype(p),xyl) / htop(p) * fr * (x(root)-x(xyl)-grav1)
+    !!TODO waterflux_inst%sapflow_patch = &
+    !!TODO         tsai(p)*params_inst%kmax(veg_pp%itype(p),xyl) / htop(p) * fr * (x(root)-x(xyl)-grav1)
 
```
waterflux_vars%sapflow_patch isn't used anywhere else in ELM as far as I can tell. (grep -rin sapflow_patch only matches that line and two instances in WaterfluxType )
  But, this was a relatively recent change (after variables like veg_wf had been added)  when the hydraulic stress was added in.  So I put the TODO to check to see if this is a mistake with the HydraulicStress functions or if it's used in some external_model I'm unaware of.

thorntonpe replied:
Got it, thanks.

peterdschwartz commented on components/elm/src/biogeophys/LakeHydrologyMod.F90:
```diff
@@ -674,13 +687,12 @@ subroutine LakeHydrology(bounds, &
          qflx_irrig(p)         = 0._r8
          qflx_irrig_col(c)     = 0._r8
 
-         ! Insure water balance using qflx_qrgwl
+         ! Insure water balance using qflx_qrgw
```
Fixed.

rljacob commented:
@sarats I added you as a reviewer not to do a line-by-line review, but just to have someone from the PG check the performance implications.

sarats commented:
@peterdschwartz Please provide a summary of timing comparison focusing on ELM timers before and after this cleanup exercise. It would be great to document if there was an improvement. We can also take a closer look at individual timers If there is a degradation (>10%).

rgknox commented:
@peterdschwartz : "Some modules have been changed to public by default to allow variables to be copied over to device in separate modules."  I don't have strong feelings about this, but I do know that enforcing private, in a general sense, is slightly more desirable from a scoping and protections perspective.  If enabling public makes your changes here more straightforward and cleaner though, I have no issues with it.

bishtgautam commented on components/elm/src/biogeochem/CarbonIsoFluxMod.F90:
```diff
@@ -918,40 +918,41 @@ end subroutine CarbonIsoFlux3
     integer                , intent(in)    :: filter_soilc(:) ! filter for soil columns
     type(cnstate_type)     , intent(in)    :: cnstate_vars
-    type(carbonflux_type)  , intent(inout) :: carbonflux_vars
+    !type(carbonflux_type)  , intent(inout) :: carbonflux_vars
```
see previous comment

bishtgautam commented on components/elm/src/biogeochem/CarbonIsoFluxMod.F90:
```diff
@@ -993,67 +994,68 @@ end subroutine CNCIsoLitterToColumn
      integer               , intent(in)    :: filter_soilc(:)   ! soil column filter
      type(cnstate_type)    , intent(in)    :: cnstate_vars
-     type(carbonflux_type) , intent(inout) :: carbonflux_vars
+     !type(carbonflux_type) , intent(inout) :: carbonflux_vars
```
same as before

bishtgautam commented on components/elm/src/biogeochem/CarbonIsoFluxMod.F90:
```diff
@@ -1140,65 +1142,66 @@ end subroutine CNCIsoGapPftToColumn
      integer               , intent(in)    :: filter_soilc(:)   ! soil column filter
      type(cnstate_type)    , intent(in)    :: cnstate_vars
-     type(carbonflux_type) , intent(inout) :: carbonflux_vars
+     !type(carbonflux_type) , intent(inout) :: carbonflux_vars
```
same

bishtgautam commented on components/elm/src/biogeochem/CarbonStateUpdate1Mod.F90:
```diff
@@ -145,26 +137,27 @@ subroutine CarbonStateUpdateDynPatch(bounds, num_soilc_with_inactive, &
 
     ! set time steps
-    dt = real( get_step_size(), r8 )
+    !#py dt = real( get_step_size(), r8 )
```
same as before

bishtgautam commented on components/elm/src/biogeochem/CropHarvestPoolsMod.F90:
```diff
@@ -87,20 +69,20 @@ subroutine CropHarvestPools(num_soilc, filter_soilc, &
 
     ! set time steps
-    dt = real( get_step_size(), r8 )
+    !#py dt = real( get_step_size(), r8 )
```
same as before

bishtgautam commented on components/elm/src/biogeochem/DUSTMod.F90:
```diff
@@ -718,14 +727,14 @@ subroutine InitDustVars(this, bounds)
-         write(iulog,*) 'dstmbl: ryn_nbr_frc_thr_prx_opt < 0.03'
-         call endrun(msg=errMsg(__FILE__, __LINE__))
+          write(iulog,*) 'dstmbl: ryn_nbr_frc_thr_prx_opt < 0.03'
+         !#py !#py call endrun(msg=errMsg(__FILE__, __LINE__))
```
same as before

bishtgautam commented on components/elm/src/biogeochem/DUSTMod.F90:
```diff
@@ -886,25 +898,27 @@ subroutine InitDustVars(this, bounds)
             if (itr_idx > 20) then
-               write(iulog,*) 'Dustini error: Terminal velocity not converging ',&
-                    ' in stk_crc_get(), breaking loop...'
+               !#py write(iulog,*) 'Dustini error: Terminal velocity not converging ',&
```
same as before

bishtgautam commented on components/elm/src/biogeochem/DryDepVelocity.F90:
```diff
@@ -252,20 +253,20 @@ subroutine depvel_compute( bounds, &
 
-            !map ELM veg type into Wesely veg type  
-            wesveg = wveg_unset 
+            !map CLM veg type into Wesely veg type
```
Revert the change to be "ELM"

bishtgautam commented on components/elm/src/biogeochem/DryDepVelocity.F90:
```diff
@@ -331,35 +333,35 @@ subroutine depvel_compute( bounds, &
-            if (index_season<0) then 
-               call endrun('ERROR: not able to determine season'//errmsg(__FILE__, __LINE__))
+            if (index_season<0) then
+               !#py call endrun('ERROR: not able to determine season'//errmsg(__FILE__, __LINE__))
```
same as before

bishtgautam commented on components/elm/src/biogeochem/EcosystemDynMod.F90:
```diff
@@ -144,115 +144,106 @@ subroutine EcosystemDynLeaching(bounds, num_soilc, filter_soilc, &
-    
-    call t_startf('PhosphorusAdsportion')
+         cnstate_vars, dt)
+    call t_stopGPU(event)
```
`t_*GPU` is a misleading name as the subroutine is applicable for CPU and GPU. An alterante may be `t_*_lnd`. @sarats may have better suggestions.

sarats replied:
Not sure what t_stopGPU does. Could you clarify? Are you doing GPU profiling through CUPTI? Is there a missing start('PhosphorusAdsportion') or just moved elsewhere?

peterdschwartz replied:
The idea for these timers was to be a simple wrapper to be used by anyone doing GPU porting to leave the CPU timers unchanged.  I thought it might be better to eventually include a routine like this in perfMod itself for any future E3SM gpu development work, which is why I used the generic *GPU name.  But, I'm happy renaming it to whatever fits best.

For profiling, I have been using nvprof and then switched to Nsight.  I do think that adding API calls or adding custom device timing routines to this timer is a great idea.  The profiling in E3SM as a whole is very messy especially and given module/routine level information is unreliable since nvprof has been deprecated.  Unit testing is helpful but it doesn't provide an accurate picture of which routines are bottlenecks.

bishtgautam commented on components/elm/src/biogeochem/EcosystemDynMod.F90:
```diff
@@ -575,382 +517,350 @@ subroutine EcosystemDynNoLeaching2(bounds,                                  &
 
        ! Transfer fates litter fluxes into ELM source arrays
-       call elm_fates%UpdateLitterFluxes(bounds)
+       call alm_fates%UpdateLitterFluxes(bounds)
```
please change it back to `elm_fates`

cc: @rgknox

peterdschwartz replied:
There's more instances of changing alm_fates from being passed an argument to using it directly through elm_instMod.  This way #ifndef _OPENACC flags can be used around these subroutine calls and the alm_fates data type doesn't need a device copy.

bishtgautam commented on components/elm/src/biogeochem/ErosionMod.F90:
```diff
@@ -1,69 +1,56 @@
-  use shr_log_mod       , only : errMsg => shr_log_errMsg
   use decompMod         , only : bounds_type
-  use clm_time_manager  , only : get_step_size
+  !#py use clm_time_manager  , only : get_step_size
```
delete the line

bishtgautam commented on components/elm/src/biogeochem/ErosionMod.F90:
```diff
@@ -112,22 +98,22 @@ subroutine ErosionFluxes (bounds, num_soilc, filter_soilc, &
          do fc = 1, num_soilc
             c = filter_soilc(fc)
 
+            print *, "fc, c:", fc, c
```
delete the line

bishtgautam commented on components/elm/src/biogeochem/FireMod.F90:
```diff
@@ -372,7 +370,7 @@ subroutine FireArea (bounds, &
               ! For non-crop -- natural vegetation and bare-soil
               if( veg_pp%itype(p)  <  nc3crop .and. cropf_col(c)  <  1.0_r8 )then
-                 if( .not. shr_infnan_isnan(btran2(p))) then
+                 if( .not. (btran2(p) .ne. btran2(p)))then !?shr_infnan_isnan(btran2(p))) then
```
I can't figure out his change equivalent to the original code.

peterdschwartz replied:
As far as I know, support for functions like is_nan() is a work-in-progess for gpu device code. This should be a platform independent way to check for NaNs.

(x .ne . x) will only ever be true for a NaN.  and so should be equivalent to shr_infnan_isnan(x).  And then in this case, it is negated to get only not NaNs.     I think this is currently the only instance where NaNs are checked for in the elm_run code

bishtgautam commented on components/elm/src/biogeochem/FireMod.F90:
```diff
@@ -1476,7 +1419,7 @@ subroutine FireFluxes (num_soilc, filter_soilc, num_soilp, filter_soilp, &
 
      if (transient_landcover) then    !true when landuse data is used
-        call get_curr_date (kyr, kmo, kda, mcsec)
+        !#py call get_curr_date (kyr, kmo, kda, mcsec)
```
delete the line

bishtgautam commented on components/elm/src/biogeochem/NitrogenStateUpdate1Mod.F90:
```diff
@@ -103,51 +101,51 @@ end subroutine NitrogenStateUpdateDynPatch
 
       ! set time steps
-      dt = real( get_step_size(), r8 )
+      !#py dt = real( get_step_size(), r8 )
```
delete the line

bishtgautam commented on components/elm/src/biogeochem/NitrogenStateUpdate1Mod.F90:
```diff
@@ -271,75 +269,75 @@ subroutine NitrogenStateUpdate1(num_soilc, filter_soilc, num_soilp, filter_soilp
 
       ! forest fertilization
-      call get_curr_date(kyr, kmo, kda, mcsec)
+      !#py call get_curr_date(kyr, kmo, kda, mcsec)
```
delete

bishtgautam commented on components/elm/src/biogeochem/NitrogenStateUpdate2Mod.F90:
```diff
@@ -6,7 +6,7 @@ module NitrogenStateUpdate2Mod
   ! !USES:
   use shr_kind_mod        , only : r8 => shr_kind_r8
-  use clm_time_manager    , only : get_step_size
+  !#py use clm_time_manager    , only : get_step_size
```
delete

bishtgautam commented on components/elm/src/biogeochem/NitrogenStateUpdate2Mod.F90:
```diff
@@ -31,37 +31,33 @@ module NitrogenStateUpdate2Mod
     integer                  , intent(in)    :: filter_soilp(:) ! filter for soil patches
-    type(nitrogenflux_type)  , intent(inout) :: nitrogenflux_vars
-    type(nitrogenstate_type) , intent(inout) :: nitrogenstate_vars
+    !type(nitrogenflux_type)  , intent(inout) :: nitrogenflux_vars
+    !type(nitrogenstate_type) , intent(inout) :: nitrogenstate_vars
```
delete

bishtgautam commented on components/elm/src/biogeochem/NitrogenStateUpdate2Mod.F90:
```diff
@@ -31,37 +31,33 @@ module NitrogenStateUpdate2Mod
-
       ! set time steps
-      dt = real( get_step_size(), r8 )
+      !#py dt = real( get_step_size(), r8 )
```
There are too many lines that are commented out. I'm now not going to add a comment on each line hat is commented out in this first review.

bishtgautam commented on components/elm/src/biogeochem/ComputeSeedMod.F90:
```diff
@@ -269,8 +245,7 @@ function SpeciesTypeMultiplier(species, pft_type, component) result(multiplier)
        case default
-          write(iulog,*) subname//' ERROR: unknown component: ', component
-          call endrun(subname//': unknown component')
+          print *, ' ERROR: unknown component: ', component
```
Please undo the deletion of all such code in this PR and wrap it with `#ifndef _OPENACC`

```
#ifndef _OPENACC
          write(iulog,*) subname//' ERROR: unknown component: ', component
          call endrun(subname//': unknown component')
#endif
```

bishtgautam reviewed (changes requested):
There is too big a PR for me to review it in one go. Thus, I have reviewed it partially, and here are my review comments.

peterdschwartz commented on components/elm/src/biogeochem/AnnualUpdateMod.F90:
```diff
@@ -25,90 +26,103 @@ module AnnualUpdateMod
     integer :: c,p          ! indices
     integer :: fp,fc        ! lake filter indices
-    real(r8):: dt           ! radiation time step (seconds)
+    real(r8) :: dayspyr = 365.d0
```
Fixed to use daypyr from the timeinfoMod.

peterdschwartz commented on components/elm/src/biogeochem/C14DecayMod.F90:
```diff
@@ -163,132 +167,135 @@ subroutine C14Decay( num_soilc, filter_soilc, num_soilp, filter_soilp, &
+              call endrun(msg=errMsg(__FILE__, __LINE__))
+           endif
+        end do
+     endif
```
Should be resolved

peterdschwartz commented:
> @peterdschwartz Please provide a summary of timing comparison focusing on ELM timers before and after this cleanup exercise. It would be great to document if there was an improvement. We can also take a closer look at individual timers If there is a degradation (>10%).

@sarats Just letting you know I'm working on this but I noticed messages in the log files like "newchild: child CPL:RUN_LOOP can't be a parent of itself"  that I'm trying resolve.

rgknox commented:
@peterdschwartz I don't have any further comments right now, I will finish the review once the conflicts are resolved.

sarats commented:
> @sarats Just letting you know I'm working on this but I noticed messages in the log files like "newchild: child CPL:RUN_LOOP can't be a parent of itself" that I'm trying resolve.

Could you double check that any additional timers you added are being closed properly, like matching start/stop? Also clarify regarding t_stopGPU.

sarats commented on components/elm/src/biogeochem/EcosystemDynMod.F90:
```diff
@@ -144,115 +144,106 @@ subroutine EcosystemDynLeaching(bounds, num_soilc, filter_soilc, &
-    
-    call t_startf('PhosphorusDesoprtion')
+         cnstate_vars, dt)
+    call t_stopGPU(event)
```
t_startGPU/t_stopGPU(event) are easy to reuse. However, it's unclear what scope they refer to. So, perhaps include the 'timer_name' in the call invocations. I guess this would probably reveal a missing t_startf/t_stopf that's causing issues.

sarats replied:
It looks like you removed the timers on the CPU side. Would timing info be absent when someone runs this just on the CPU?

sarats replied:
@peterdschwartz Revisiting this topic of timers: It's still unclear how t_stopGPU is implemented and how it handles CPU only runs. 

Are these very fine-grained timers for GPU? If so, you are probably better off with nvprof/nsys style tool. The GPTL timers are primarily meant for application phases.

peterdschwartz commented:
@rgknox re: public vs private.   After checking and doing some tests, the only modules I have to blanket change to public for this PR are ColumnDataType and VegetationDataType because accessing the type-bound procedures doesn't seem possible with private default even though they are declared public in the type (compiler bug or edge-case not covered by Fortran standard?).   These modules seem to already have everything labeled public except for some procedures that are explicitly made private, so I don't think getting rid of the private setting here would have consequences.  Do you agree?   
The individual variables from other modules that needed to be made public have been done so on an individual basis rather than making the whole module public.  It's possible to maintain them being private and copying to device in the readNML routines but that would be suboptimal use of bandwidth, so I'd prefer to keep them public.

rgknox commented:
I'm in agreement @peterdschwartz

peterdschwartz commented:
I rebased this branch to the master head as of Tuesday May 4th and did baseline comparisons for e3sm_land_developer tests that all passed.  I squashed several of the smaller commits and made new changes to address the review comments as well.

sarats commented on components/elm/src/main/perfMod_GPU.F90:
```diff
@@ -7,25 +7,25 @@ module perfMod_GPU
 
 contains
 
-  subroutine  t_startGPU(event)
```
So, you fallback to plain GPTL calls (t_startf) when there is no OpenACC. And it's presently just a placeholder for the GPU side timing code? What are you planning to use?

peterdschwartz replied:
Yes it just a placeholder to avoid having to comment out all the timing functions for cpu runs.  A future development I'm interested in would be to make calls to the CUDA  clock() device function to get rough timing for sections of more complicated kernels on a GPU block level.  

 For profiling purposes, I currently either use the built-in PGI_ACC_TIME (which is CUPTI based)  for running the full ELM code or use nvprof/nv_nsight_cu_cli to profile certain combinations of subroutines in a unit-testing environment.

Also, I have the timing data comparing this branch to master for a few runs on Summit.  The compset I used was I1850GSWCNPRDCTCBC with hcru_hcru resolution with ~500 gridcells per process.  The run length was 31 days and I just compared the model_timing_stats times.   It seems that this branch is consistently a little bit faster.  The "CNUpdate2" is nearly twice as fast. 

The times on summit seem pretty consistent and here are the Walltotal times for a set of runs.  Let me know if you want the files or a different comparison.   I include the dyn_subgrid since this branch made many changes to that section.

**Timer**            **master**         **branch**
l:elm_run            9.97e+04           9.48e+04
l:dyn_subgrid     7.86e+03           7.70e+03
l:CNUpdate2      5.46e+03           2.70e+03

peterdschwartz replied:
@sarats Does anything else need to be resolved concerning the timers or for performance?

sarats replied:
Timings look good. 

The placement of perfMod_GPU.F90 under the land model sources still feels a bit out of place. As it is, it is a stub on both CPU and GPU fronts.  I'm Ok with moving forward for now but we should probably converge on requirements if there is a need for fine-granularity of timers within GPU code.

sarats replied:
@peterdschwartz I approved this PR but there is still a need to converge on GPU timer requirements for the future.

rgknox commented:
FYI: @peterdschwartz and I have been coordinating on code changes, as we both have made significant modifications to AllocationMod.  I'm in the process of submitting a PR that is based on his (this) set of changes.  One thing that I'm a little concerned about is how a rebase on this branch will effect my process.  

For those interested, here is the diff:

https://github.com/E3SM-Project/E3SM/compare/peterdschwartz/lnd/elm_cleanup_openacc_prep...rgknox:rgknox/lnd/fates-cnp-openaccbase?expand=1

@rljacob @bishtgautam

rljacob commented:
@rgknox are you saying you made a branch off of this branch and you wonder what happens if this branch gets rebased?

rgknox commented:
@rljacob , yes exactly

rljacob commented:
I *think* your branch will also move but I'm not sure.  You may need to rebase it on to the equivalent point on the new branch.

rljacob commented:
@peterdschwartz please edit the title to remove the branch name.

bishtgautam commented:
Please rebase this branch.

peterdschwartz commented:
Well I rebased and tested the branch but it seems a PR went in that modified several parts of the ELM code (PR 4229?).  The conflict for UrbanFluxesMod.F90 produced by git is messy, so the branch likely needs another rebase.

peterdschwartz commented:
@bishtgautam All the tests I ran for this rebase passed and no more conflicts, so I think everything should be good to go.

dqwu commented:
@peterdschwartz 
For ne4 F case on anlworkstation with GNU compiler, I got some build errors with latest next branch:
```
components/elm/src/biogeochem/DUSTMod.F90:113:105:

     allocate(this%flx_mss_vrt_dst_patch     (begp:endp,1:ndst)) ; this%flx_mss_vrt_dst_patch     (:,:) = nan
                                                                                                         1
Error: Can't convert TYPE(shr_infnan_nan_type) to REAL(8) at (1)
```
I think it is related to this PR.

peterdschwartz commented:
@dqwu Thanks for letting me know. I have fixed the issue and ran the test successfully with gnu.
